mobly/controllers/android_device_lib/snippet_client_v2.py

```diff
@@ -322,27 +613,72 @@ def _stop_server(self):
           self._device,
           f'Failed to stop existing apk. Unexpected output: {out}.')
 
-  # TODO(mhaoli): Temporally override these abstract methods so that we can
-  # initialize the instances in unit tests. We are implementing these functions
-  # in the next PR as soon as possible.
-  def make_connection(self):
-    raise NotImplementedError('To be implemented.')
+  def restore_server_connection(self, port=None):
+    """Restores the server after the device got reconnected.
 
-  def close_connection(self):
-    raise NotImplementedError('To be implemented.')
+    Instead of creating a new instance of the client:
+      - Uses the given port (or find a new available host_port if none is
+      given).
+      - Tries to connect to the remote server with the selected port.
 
-  def __del__(self):
-    # Override the destructor to not call close_connection for now.
-    pass
+    Args:
+      port: int, if given, this is the host port from which to connect to the
+        remote device port. If not provided, find a new available port as host
+        port.
 
-  def send_rpc_request(self, request):
-    raise NotImplementedError('To be implemented.')
+    Raises:
+      errors.ServerRestoreConnectionError: when failed to restore the connection
+        to the snippet server.
+    """
+    try:
+      # If self.host_port is None, self._make_connection finds a new available
+      # port.
+      self.host_port = port
+      self._make_connection()
+    except Exception as e:
+      # Log the original error and raise ServerRestoreConnectionError.
+      self.log.error('Failed to re-connect to the server.')
+      raise errors.ServerRestoreConnectionError(
+          self._device,
+          (f'Failed to restore server connection for {self.package} at '
+           f'host port {self.host_port}, device port {self.device_port}.'
+          )) from e
 
-  def check_server_proc_running(self):
-    raise NotImplementedError('To be implemented.')
+    # Because the previous connection was lost, update self._proc
+    self._proc = None
+    self._restore_event_client()
 
-  def handle_callback(self, callback_id, ret_value, rpc_func_name):
-    raise NotImplementedError('To be implemented.')
+  def _restore_event_client(self):
+    """Restores the previously created event client or creates a new one.
 
-  def restore_server_connection(self, port=None):
-    raise NotImplementedError('To be implemented.')
+    This function restores the connection of the previously created event
+    client, or creates a new client and makes a connection if it didn't
+    exist before.
+
+    The event client to restore reuses the same host port and device port
+    with the client on which function is called.
+    """
+    if not self._event_client:
```

Comment:
Not clear why there is a branching here.
In which situations would `_event_client` be None or not None?

Reply:
Here I followed the same branch as in client v1. `self._event_client` is None until the snippet client has sent the first async RPC. After that, `self._event_client` won't be None.

Reply:
If the `_event_client` is not even initialized, why do you need to create it in restoration? shouldn't the restoration just be no-op?

Reply:
Removed this branch. 

It seems more reasonable to be no-op. Also checked the git PR which introduced this logic and didn't find any reason to create it here.